File: app/src/main/java/com/friendlypos/distribucion/fragment/DistSelecProductoFragment.java
package com.friendlypos.distribucion.fragment;

import android.app.Activity;
import android.os.Bundle;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.friendlypos.R;
import com.friendlypos.Recibos.modelo.receipts;
import com.friendlypos.Recibos.modelo.recibos;
import com.friendlypos.distribucion.activity.DistribucionActivity;
import com.friendlypos.distribucion.adapters.DistrResumenAdapter;
import com.friendlypos.distribucion.adapters.DistrSeleccionarProductosAdapter;
import com.friendlypos.distribucion.modelo.Inventario;
import com.friendlypos.distribucion.util.TotalizeHelper;
import com.friendlypos.principal.modelo.Clientes;
import com.friendlypos.principal.modelo.Productos;

import java.util.ArrayList;
import java.util.List;
import java.util.zip.Inflater;

import io.realm.Realm;
import io.realm.RealmQuery;
import io.realm.RealmResults;

import static io.realm.internal.SyncObjectServerFacade.getApplicationContext;


public class DistSelecProductoFragment extends BaseFragment implements SearchView.OnQueryTextListener {
    private Realm realm;
    RecyclerView recyclerView;
    private DistrSeleccionarProductosAdapter adapter;
    private DistrResumenAdapter adapter2;
    private DistResumenFragment resumenFrag1;
    private static int bill_type = 1;
    static TextView creditoLimite;
    static double creditoLimiteCliente = 0.0;
    int slecTAB;
    DistribucionActivity activity;
    TotalizeHelper totalizeHelper;
    int datosEnFiltro=0;

    public static DistSelecProductoFragment getInstance() {
        return new DistSelecProductoFragment();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Override
    public void onAttach(Activity activity){
        super.onAttach(activity);
        this.activity = (DistribucionActivity) activity;
    }

    @Override
    public void onDetach(){
        super.onDetach();
        activity = null;

    }
    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment

        View rootView = inflater.inflate(R.layout.fragment_distribucion_selecproduct, container,
            false);
        setHasOptionsMenu(true);
        recyclerView = (RecyclerView) rootView.findViewById(R.id.recyclerViewDistrSeleccProducto);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerView.setHasFixedSize(true);

        if(adapter == null) {
            adapter = new DistrSeleccionarProductosAdapter(activity, this, getListProductos());
        }
        recyclerView.setAdapter(adapter);
        creditoLimite = (TextView) rootView.findViewById(R.id.restCredit);

        Log.d("listaProducto", getListProductos() + "");
        adapter2 = new DistrResumenAdapter();
        creditoDisponible();

        return rootView;

    }

    private List<Inventario> getListProductos() {
        realm = Realm.getDefaultInstance();


                RealmQuery<Inventario> query1 = realm.where(Inventario.class).notEqualTo("amount", "0").notEqualTo("amount", "0.0").notEqualTo("amount", "0.000");
                RealmResults<Inventario> result2 = query1.findAll();

        if (result2.isEmpty()) {

            Toast.makeText(getApplicationContext(),"Favor descargar datos primero",Toast.LENGTH_LONG).show();}

        else{
            for (int i = 0; i < result2.size(); i++) {

                List<Inventario> salesList1 = realm.where(Inventario.class).notEqualTo("amount", "0").notEqualTo("amount", "0.0").notEqualTo("amount", "0.000").findAll();
                String nombre = salesList1.get(i).getNombre_producto();
                if (nombre == null){
                    String facturaId1 = salesList1.get(i).getProduct_id();

                    Productos salesList2 = realm.where(Productos.class).equalTo("id", facturaId1).findFirst();

                    final String facturaId2 = salesList2.getId();
                    final String desc = salesList2.getDescription();

                    final Realm realm3 = Realm.getDefaultInstance();

                    realm3.executeTransactionAsync(new Realm.Transaction() {
                        @Override
                        public void execute(Realm bgRealm) {
                            Inventario inv_actualizado = realm3.where(Inventario.class).equalTo("product_id", facturaId2).findFirst();
                            //  inv_actualizado.setProducto(new RealmList<Productos>(salesList2.toArray(new Productos[salesList2.size()])));
                            inv_actualizado.setNombre_producto(desc);
                            realm3.insertOrUpdate(inv_actualizado); // using insert API
                        }
                    }, new Realm.Transaction.OnSuccess() {
                        @Override
                        public void onSuccess() {
                            realm3.close();
                        }
                    }, new Realm.Transaction.OnError() {
                        @Override
                        public void onError(Throwable error) {
                            realm3.close();
                        }
                    });


                }
            }

        }

        return result2;

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        realm.close();
    }

    public void creditoDisponible() {
        slecTAB = activity.getSelecClienteTab();

        if (slecTAB == 1){

        String metodoPagoCliente = activity.getMetodoPagoCliente();
        String dueCliente = activity.getDueCliente();

        Log.d("PagoProductoSelec", metodoPagoCliente + "");
        Log.d("PagoProductoSelec", creditoLimiteCliente + "");
        Log.d("PagoProductoSelec", dueCliente + "");

        if (metodoPagoCliente.equals("1")) {
            bill_type = 1;
            creditoLimite.setVisibility(View.GONE);
        }
        else if (metodoPagoCliente.equals("2")) {
            bill_type = 2;
            try {
                creditoLimite.setVisibility(View.VISIBLE);
                creditoLimite.setText("C.Disponible: " + String.format("%,.2f", creditoLimiteCliente));
            }
            catch (Exception e) {
                Log.d("JD", "Error " + e.getMessage());
            }
        }}
        else{
            Toast.makeText(getActivity(),"nadaSelecProducto",Toast.LENGTH_LONG).show();
        }

    }

    @Override
    public void updateData() {


        if(datosEnFiltro == 1){
            Log.d("OSCARUpdate", "No actualiza xq esta en " + datosEnFiltro);
        }
        else{
            datosEnFiltro = 0;
            adapter.updateData(getListProductos());
            adapter2.notifyDataSetChanged();
            Log.d("OSCARUpdate1", "Actualiza xq esta en " + datosEnFiltro);
        }


        if (slecTAB == 1) {
            creditoLimiteCliente = Double.parseDouble(((DistribucionActivity) getActivity()).getCreditoLimiteCliente());
            creditoLimite.setText("C.Disponible: " + String.format("%,.2f", creditoLimiteCliente));
        }
        else{
            Toast.makeText(getActivity(),"nadaSelecProducto",Toast.LENGTH_LONG).show();
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_main, menu);

        final MenuItem item = menu.findItem(R.id.action_search);
        final SearchView searchView = (SearchView) MenuItemCompat.getActionView(item);
        searchView.setOnQueryTextListener(this);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        final List<Inventario> filteredModelList = filter(getListProductos(), newText);
        adapter.setFilter(filteredModelList);
        return true;
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }


    private List<Inventario> filter(List<Inventario> models, String query) {


        if(query.isEmpty()){
            Log.d("OSCARVAC", "esta vacio la consulta");
            datosEnFiltro = 0;

        }else{

            datosEnFiltro = 1;

            Log.d("OSCARLLE", "esta llena la consulta");
        }

        query = query.toLowerCase();

        final List<Inventario> filteredModelList = new ArrayList<>();
        for (Inventario model : models) {

                if(model.getNombre_producto() == null){

                }else{

            final String text = model.getNombre_producto().toLowerCase();
            Log.d("dasdad", text);
            if (text.contains(query)) {
                filteredModelList.add(model);
            }}
        }
        return filteredModelList;
    }


}

File: app/src/main/java/com/friendlypos/distribucion/fragment/DistSelecProductoFragment.kt
package com.friendlysystemgroup.friendlypos.distribucion.fragment

import android.app.Activity
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import androidx.core.view.MenuItemCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.friendlysystemgroup.friendlypos.R
import com.friendlysystemgroup.friendlypos.databinding.FragmentDistSelecProductoBinding
import com.friendlysystemgroup.friendlypos.distribucion.activity.DistribucionActivity
import com.friendlysystemgroup.friendlypos.distribucion.adapters.DistrResumenAdapter
import com.friendlysystemgroup.friendlypos.distribucion.adapters.DistrSeleccionarProductosAdapter
import com.friendlysystemgroup.friendlypos.distribucion.modelo.Inventario
import com.friendlysystemgroup.friendlypos.distribucion.util.TotalizeHelper
import com.friendlysystemgroup.friendlypos.modelos.Producto
import io.realm.Realm
import io.realm.RealmResults
import java.util.Locale

class DistSelecProductoFragment : BaseFragment() {
    private var binding: FragmentDistSelecProductoBinding? = null
    private var realm: Realm? = null
    var recyclerView: RecyclerView? = null
    private var adapter: DistrSeleccionarProductosAdapter? = null
    private var adapter2: DistrResumenAdapter? = null
    private val resumenFrag1: DistResumenFragment? = null
    var slecTAB: Int = 0
    var activity: DistribucionActivity? = null
    var totalizeHelper: TotalizeHelper? = null
    var datosEnFiltro: Int = 0
    private var productos: List<Producto> = listOf()
    private val productosFilter: MutableList<Producto> = mutableListOf()

    @Suppress("DEPRECATION")
    override fun onAttach(activity: Activity) {
        super.onAttach(activity)
        this.activity = activity as DistribucionActivity
    }

    override fun onDetach() {
        super.onDetach()
        activity = null
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentDistSelecProductoBinding.inflate(inflater, container, false)
        return binding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        
        setupViews()
    }

    private fun setupViews() {
        binding?.apply {
            recyclerViewDistProductos?.layoutManager = LinearLayoutManager(context)
            
            productos = listaProductos
            productosFilter.clear()
            productosFilter.addAll(productos)
            
            adapter = DistrSeleccionarProductosAdapter(
                context,
                this@DistSelecProductoFragment.activity,
                productosFilter
            )
            recyclerViewDistProductos?.adapter = adapter
            
            searchView?.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String): Boolean {
                    return false
                }

                override fun onQueryTextChange(newText: String): Boolean {
                    filterList(newText)
                    return true
                }
            })
        }
    }

    private fun filterList(filterText: String) {
        productosFilter.clear()
        
        if (filterText.isEmpty()) {
            productosFilter.addAll(productos)
        } else {
            val lowerCaseFilter = filterText.lowercase(Locale.getDefault())
            
            productos.forEach { producto ->
                val nombre = producto.nombre?.lowercase(Locale.getDefault()) ?: ""
                val codigo = producto.codigo?.lowercase(Locale.getDefault()) ?: ""
                
                if (nombre.contains(lowerCaseFilter) || codigo.contains(lowerCaseFilter)) {
                    productosFilter.add(producto)
                }
            }
        }
        
        adapter?.notifyDataSetChanged()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
        realm?.close()
    }

    fun creditoDisponible() {
        slecTAB = activity!!.selecClienteTab

        if (slecTAB == 1) {
            val metodoPagoCliente = activity!!.metodoPagoCliente
            val dueCliente = activity!!.dueCliente

            Log.d("PagoProductoSelec", metodoPagoCliente + "")
            Log.d("PagoProductoSelec", creditoLimiteCliente.toString() + "")
            Log.d("PagoProductoSelec", dueCliente + "")

            if (metodoPagoCliente == "1") {
                bill_type = 1
                creditoLimite!!.visibility =
                    View.GONE
            } else if (metodoPagoCliente == "2") {
                bill_type = 2
                try {
                    creditoLimite!!.visibility =
                        View.VISIBLE
                    creditoLimite!!.text =
                        "C.Disponible: " + String.format(
                            "%,.2f",
                            creditoLimiteCliente
                        )
                } catch (e: Exception) {
                    Log.d("JD", "Error " + e.message)
                }
            }
        } else {
            Toast.makeText(getActivity(), "nadaSelecProducto", Toast.LENGTH_LONG).show()
        }
    }

    override fun updateData() {
        if (datosEnFiltro == 1) {
            Log.d("OSCARUpdate", "No actualiza xq esta en $datosEnFiltro")
        } else {
            datosEnFiltro = 0
            adapter!!.updateData(listProductos)
            adapter2!!.notifyDataSetChanged()
            Log.d("OSCARUpdate1", "Actualiza xq esta en $datosEnFiltro")
        }


        if (slecTAB == 1) {
            creditoLimiteCliente =
                (getActivity() as DistribucionActivity).creditoLimiteCliente!!.toDouble()
            creditoLimite!!.text =
                "C.Disponible: " + String.format(
                    "%,.2f",
                    creditoLimiteCliente
                )
        } else {
            Toast.makeText(getActivity(), "nadaSelecProducto", Toast.LENGTH_LONG).show()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_main, menu)

        val item = menu.findItem(R.id.action_search)
        val searchView = MenuItemCompat.getActionView(item) as SearchView
        searchView.setOnQueryTextListener(this)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onQueryTextChange(newText: String): Boolean {
        val filteredModelList = filter(listProductos, newText)
        adapter!!.setFilter(filteredModelList)
        return true
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        return false
    }


    private fun filter(models: List<Inventario>, query: String): List<Inventario> {
        var query = query
        if (query.isEmpty()) {
            Log.d("OSCARVAC", "esta vacio la consulta")
            datosEnFiltro = 0
        } else {
            datosEnFiltro = 1

            Log.d("OSCARLLE", "esta llena la consulta")
        }

        query = query.lowercase(Locale.getDefault())

        val filteredModelList: MutableList<Inventario> = ArrayList()
        for (model in models) {
            if (model.description == null) {
            } else {
                val text = model.description.lowercase(Locale.getDefault())
                Log.d("dasdad", text)
                if (text.contains(query)) {
                    filteredModelList.add(model)
                }
            }
        }
        return filteredModelList
    }

    val listaProductos: List<Producto>
        get() {
            val realm = Realm.getDefaultInstance()
            try {
                val productos: RealmResults<Producto> = realm.where(Producto::class.java)
                    .equalTo("isVisible", true)
                    .findAll()
                
                return realm.copyFromRealm(productos)
            } finally {
                realm.close()
            }
        }

    companion object {
        private var bill_type = 1
        var creditoLimite: TextView? = null
        var creditoLimiteCliente: Double = 0.0
        val instance: DistSelecProductoFragment
            get() = DistSelecProductoFragment()
    }
}
